pkg/parser/grammar/where_clause_visitor.go

```diff
@@ -0,0 +1,587 @@
+		case "hasAll":
+			cond = fmt.Sprintf("hasAll(%s, %s)", fieldName, v.builder.Var(value))
+		}
+		conds = append(conds, cond)
+	}
+
+	return v.builder.Or(conds...)
+}
+
+// VisitFunctionParamList handles the parameter list for function calls
+func (v *WhereClauseVisitor) VisitFunctionParamList(ctx *FunctionParamListContext) any {
+	params := ctx.AllFunctionParam()
+	parts := make([]any, len(params))
+
+	for i, param := range params {
+		parts[i] = v.Visit(param)
+	}
+
+	return parts
+}
+
+// VisitFunctionParam handles individual parameters in function calls
+func (v *WhereClauseVisitor) VisitFunctionParam(ctx *FunctionParamContext) any {
+	if ctx.Key() != nil {
+		return v.Visit(ctx.Key())
+	} else if ctx.Value() != nil {
+		return v.Visit(ctx.Value())
+	} else if ctx.Array() != nil {
+		return v.Visit(ctx.Array())
+	}
+
+	return "" // Should not happen with valid input
+}
+
+// VisitArray handles array literals
+func (v *WhereClauseVisitor) VisitArray(ctx *ArrayContext) any {
+	return v.Visit(ctx.ValueList())
+}
+
+// VisitValue handles literal values: strings, numbers, booleans
+func (v *WhereClauseVisitor) VisitValue(ctx *ValueContext) any {
+	if ctx.QUOTED_TEXT() != nil {
+		txt := ctx.QUOTED_TEXT().GetText()
+		// trim quotes and return the value
+		return strings.Trim(txt, "\"'")
+	} else if ctx.NUMBER() != nil {
+		number, err := strconv.ParseFloat(ctx.NUMBER().GetText(), 64)
+		if err != nil {
+			v.errors = append(v.errors, errors.Newf(
+				errors.TypeInvalidInput,
+				errors.CodeInvalidInput,
+				"failed to parse number %s",
+				ctx.NUMBER().GetText(),
+			))
+			return ""
+		}
+		return number
+	} else if ctx.BOOL() != nil {
+		// Convert to ClickHouse boolean literal
+		boolText := strings.ToLower(ctx.BOOL().GetText())
+		return boolText == "true"
+	} else if ctx.KEY() != nil {
+		// Why do we have a KEY context here?
+		// When the user writes an expression like `service.name=redis`
+		// The `redis` part is a VALUE context but parsed as a KEY token
+		// so we return the text as is
+		return ctx.KEY().GetText()
+	}
+
+	return "" // Should not happen with valid input
+}
+
+// VisitKey handles field/column references
+func (v *WhereClauseVisitor) VisitKey(ctx *KeyContext) any {
+
+	fieldKey := telemetrytypes.GetFieldKeyFromKeyText(ctx.KEY().GetText())
+
+	keyName := strings.TrimPrefix(fieldKey.Name, telemetrylogs.BodyJSONStringSearchPrefix)
+
+	fieldKeysForName := v.fieldKeys[keyName]
+
+	// for the body json search, we need to add search on the body field even
+	// if there is a field with the same name as attribute/resource attribute
+	// Since it will ORed with the fieldKeysForName, it will not result empty
+	// when either of them have values
+	if strings.HasPrefix(fieldKey.Name, telemetrylogs.BodyJSONStringSearchPrefix) {
+		fieldKeysForName = append(fieldKeysForName, fieldKey)
+	}
+
+	// TODO(srikanthccv): do we want to return an error here?
+	// should we infer the type and auto-magically build a key for expression?
```

Comment:
this will happen when the it doesn't contain in the attribute keys  right ?

Reply:
>this will happen when the it doesn't contain in the attribute keys right ?

When the key is not present in the `fieldKeys` for some reason i.e key we have never seen or missed in keys table for any reason.

>Also where will the enrichment logic be present ?

We can't enrich a key with one type or data type because there can be more than one combination, and in that case, we query across all the combinations and return the results along with the warning. If the question is where will the `fieldKeys` come from, they will be queried from DB and passed as args.

Reply:
Hmm, so previously if we didn't find anything then we fallback to string though. While I don't have much issue with this new implementation as that is a rare case.

Reply:
Add one error test case as well, so that in future if someone changes something we know that error was expected.

Reply:
We have a retention of 15 days for keys, but the main data retention is customisable, which could potentially create a situation where keys exist in the main data but the keys don't because of key retention. The error is probably fine in a genuine error case, but in other cases, like retention, we can't return an error. I will think about what we should do about it (assuming String could result in no results when the results are expected) and address it separately.

Reply:
ok, then let's just add an error test case for now and merge this PR.

Reply:
There is already one https://github.com/SigNoz/signoz/pull/7564/files#diff-e1250b57770f50babc406cd27231128fad68f212e8705a7dbc004db6131da3faR590. See line 590 in test file.